doc/fexd.8
.\"                              hey, Emacs:   -*- nroff -*-
.\" fexd is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH FEXD 8 "November 4, 2004"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
fexd \- file exchange daemon
.SH SYNOPSIS
.B fexd
.RB [ -dvglID ]
[ \fB-c\fP \fIconfig_file\fP ]
.SH DESCRIPTION
\fBfexd\fP is a disconnected network file system very similar to
intermezzo (and not so similar to coda). In contrast to these two
\fBfexd\fP needs no kernel modules to operate. \fBfexd\fP replicates
shared files on all computers in the network.  When a disconnected
computer is reconnected \fBfexd\fP reintigrates all changes an
synchronizes the network with the reconnected computer. Unlike coda
\fBfexd\fP has no explicit authentication and authorisation mechanism.
.SS Conflicts
A conflict can arise when two disconnected sites change a file. After
reconnection two different versions of the same file exist. \fBfexd\fP
resolves conflicts by renaming one of the conflicting versions:
\fIfilename.ext\fP becomes \fIfilename-<revno>.ext\fP where
\fI<revno>\fP is the revision number of the backup starting with 1.
.SS SSH Tunneling
If two sites are connected over an unsecure line. \fBfexd\fP can be 
configured to automaticly use ssh tunneling. \fBfexd\fP uses key authentication
for connecting a ssh server. 
.SS UID and GID Translation
If the user ids and group ids differ for the same user at different sites.
\fBfexd\fP is able to translate these ids.
.SS File locking
\fBfexd\fP only supports locks on the whole file. If one site locks a file, 
the lock will be distributed over the network. If a site gets disconnected
while holding a lock the lock will be released.
.SH OPTIONS
\fBfexd\fP accepts the following options:
.TP
.BR -d ", " --debug
Do not deamonize, print messages to stdout. Repeating this 
options increases the message level.
.TP
.BR  -v ", " --verbose
Be verbose, write messages to the syslog. Repeating this options 
increases the message level.
.TP
.BR  -l ", " --no-locks
Disable lock, detection mechnism. If you do not need 
lock distribution over the network you can can turn on 
this option, because the detection mechanism is can be expensive.
.TP
.BR  -I ", " --no-inotify
Dont use the inotify mechanism to detect file changes.
.TP
.BR  -D ", " --no-dnotify
Dont use the dnotify mechanism to detect file changes.
.TP
.BR  -c ", " --config " "\fIconfigfile\fP
Read configuration information from \fIconffile\fP. 
.TP


.SH SEE ALSO
.BR fex.conf (5).

.SH AUTHOR
Michael Reithinger <mreithinger@web.de>.


NEWS
Changes 0.8.18

 * fex now compiles under gcc 3.4.4

Changes 0.8.17

 * added readonly option (for one way synchronization) to configuration.

Changes 0.8.16

 * bugfix backup mechanism: Directories may now cpntain also a period.
 * bugfix renaming directorys with inotify or dnotify works now corectly
 * inotify: added support for inotify interface after version 0.20
 

Changes 0.8.15

 * bugfix for backup mechanism: When a connection was aborted
 * during reintegration, a conflicting file was backuped 
 * multiple times.


Changes 0.8.14
 
 * create watchpoint directories if they don't exist
 * new command line options -no-inotify, -no-dnotify
 * more debug messages

Changes 0.8.13
 
 * improved net failure robustness 

Changes 0.8.12

 * improved file lock detection
 * insert help message

Changes 0.8.11

 * some bugfixes in the file lock mechanism
 * removed file copy operations

Changes 0.7.9

 * First public release.

src/fexd.cpp
/***************************************************************************
 *   Copyright (C) 2004 by Michael Reithinger                              *
 *   mreithinger@web.de                                                    *
 *                                                                         *
 *   This file is part of fex.                                             *
 *                                                                         *
 *   fex is free software; you can redistribute it and/or modify           *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   fex is distributed in the hope that it will be useful,                *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#include <log4cpp/Category.hh>
#include <log4cpp/SyslogAppender.hh>
#include <log4cpp/OstreamAppender.hh>
#include <log4cpp/SimpleLayout.hh>
#include <log4cpp/BasicLayout.hh>
#include "configfile.h"
#include "filelistener.h"
#include "connection.h"
#include <getopt.h>
#include <signal.h>
#include <sys/wait.h>


using namespace std;
using namespace nmstl;
using namespace log4cpp;

char* version_string = "fexd "PACKAGE_VERSION;
Category& lc = Category::getRoot();
io_event_loop MainLoop;
bool do_lock_polling = true;

#if defined(USE_DNOTIFY)
#define HAS_DO_INOTIFY 1
bool do_inotify = true;
#endif

#if defined(USE_POLLING)
#if ! defined(USE_DNOTIFY)
bool do_inotify = true;
#else
#define HAS_DO_DNOTIFY 1
bool do_dnotify = true;
#endif
#endif


static const char* config_file = FEX_CONF;
static int debug_level = 0;
static int verbose_level = 0;
static ofstream logstream;


static void 
parse_options(int argc, char *argv[]);

static void
prepare_log();

static
void
print_help();

static
void 
terminate(int);

static void 
check_children(int);

int
main(int argc, char *argv[])
{
  srand(time(NULL));

  parse_options(argc, argv);
  prepare_log();

  if (!debug_level) {
    if (daemon(0, 0) != 0) {
      lc.error("daemon failed! %s", strerror(errno));
      return 2;
    }
  }

  if (getuid() != 0) {
    cerr << "sorry, must be run as root" << endl;
    return 1;
  }

  lc.notice("%s started", version_string);
  Configuration::get().parse(config_file);
  ConnectionPool::get().start_listening();

  signal(SIGTERM, terminate);
  signal(SIGINT, terminate);
  signal(SIGPIPE, terminate);
  signal(SIGCHLD, check_children);

  lc.notice("daemon start to loop");
 tryagain:
  try
    {
      MainLoop.run();
    }
  catch(exception& error) 
    {
      lc.fatal(error.what());
      goto tryagain;
    }
  catch(...) {
    lc.fatal("a really bad error occured");
  }

  MainLoop.tidy_handlers();
  lc.notice("fexd finished");
  lc.shutdown();
}

  
void 
terminate(int)
{
  lc.debug("server terminated");
  MainLoop.terminate();
}

void 
check_children(int)
{
  pid_t pid = ::wait(NULL);
  if (pid > 0)
    lc.notice("child finished: %i", pid);
}


void 
parse_options(int argc, char *argv[])
{
  static struct option long_options[] = {
        { "debug"     , 0, 0, 'd' },
        { "verbose"   , 0, 0, 'v' },
	{ "no-locks"  , 0, 0, 'l' },
#ifdef HAS_DO_INOTIFY
	{ "no-inotify", 0, 0, 'I' },
#endif
#ifdef HAS_DO_DNOTIFY
	{ "no-dnotify", 0, 0, 'D' },
#endif
	{ "config"    , 0, 0, 'c' },
	{ "help"      , 0, 0, 'h' },
        { 0, 0, 0, 0 }
      };
      
  int option_index = 0;
  int arg;

  while((arg = getopt_long(argc, argv, "dvlIDch",
			   long_options, &option_index)) != -1) {
    switch(arg){
    case 'l':
      do_lock_polling = false;
      break;

    case 'd':
      debug_level++;
      break;

    case 'v':
      verbose_level++;
      break;

#ifdef HAS_DO_INOTIFY
    case 'I':
      do_inotify = false;
      break;
#endif

#ifdef HAS_DO_DNOTIFY
    case 'D':
      do_dnotify = false;
      break;
#endif

    case 'c':
      config_file = argv[optind];
      break;

    default:
      print_help();
      exit(0);
    }
  }
}

#undef WARN
#undef NOTICE
#undef INFO
#undef DEBUG

void
prepare_log()
{
  Appender* app;

  if (debug_level) {
    app = new OstreamAppender("ostream", &cerr);
    app->setLayout(new BasicLayout());
  }
  else {
    app = new SyslogAppender("syslog", "fexd", LOG_DAEMON);
    app->setLayout(new SimpleLayout());
  }

  lc.removeAllAppenders();
  lc.setAppender(app);

  switch(max(debug_level, verbose_level)) {
  case 0 : lc.setPriority(Priority::WARN  ); break;
  case 1 : lc.setPriority(Priority::NOTICE); break;
  case 2 : lc.setPriority(Priority::INFO  ); break;
  default: lc.setPriority(Priority::DEBUG ); break;
  }
}

static
void
print_help()
{
  cout << "fexd "PACKAGE_VERSION << endl;
  cout << "fexd [-d|--debug] [-v|--verbose] [-l|--no-lock ] [-h|--help]"
       << endl;

  cout << "     [-c|--config <config_file>]";
  
#ifdef HAS_DO_INOTIFY
  cout << " [-I|--no-inotify]";
#endif
#ifdef HAS_DO_DNOTIFY
  cout << " [-D|--no-dnotify]";
#endif
  cout << endl;
  cout << "  -d,--debug       don't deamonize, print debug messages to stdout"
       << endl;
  cout << "  -v,--verbose     set verbosity level" << endl;
  cout << "  -n,--no-locks    no lock detection " << endl;
  cout << "  -h,--help        print this message" << endl;

#ifdef HAS_DO_INOTIFY
  cout << "  -I,--no-inotify  don't use inotify mechanism" << endl;
#endif
#ifdef HAS_DO_DNOTIFY
  cout << "  -D,--no-dnotify  don't use dnotify mechanism" << endl;
#endif
  cout << "  -c,--config      path to alternate configuration file" << endl;
  cout << "                   (default is "FEX_CONF")\n" << endl;
}
